server/src/instant/util/aws_signature.clj

```diff
@@ -0,0 +1,270 @@
+    (if (> change-amount 0)
+      (change-directory (drop-last (rest segments)))
+      segments)))
+
+(defn- both
+  [f1 f2]
+  #(and (f1 %) (f2 %)))
+
+(defn- not-blank?
+  [str]
+  (and (not (= "" str))
+       (not (nil? str))))
+
+(defn- not-dot?
+  [str]
+  (not= str "."))
+
+(defn- resolve-path
+  [path]
+  (->> (str/split path #"/")
+       (filter (both not-blank? not-dot?))
+       (change-directory)
+       (str/join "/")
+       (str "/")))
+
+(defn- encode-uri
+  [uri]
+  (->> (str/split uri #"/")
+       (map url-encode)
+       (str/join "/")
+       (#(if (str/blank? %) "/" %))))
+
+(defn- append-slash
+  [uri raw]
+  (str uri (and (re-matches #".*/$" raw) "/")))
+
+(defn- replace-double-slash
+  [uri]
+  (str/replace uri #"//" "/"))
+
+(defn- normalize-uri
+  [uri]
+  (-> uri
+      (resolve-path)
+      (encode-uri)
+      (append-slash uri)
+      (replace-double-slash)))
+
+(defn- canonical-headers [headers]
+  (into (sorted-map)
+        (map (fn [[k v]] [(str/lower-case k) (str/trim (or v ""))]) headers)))
+
+(defn- stringify-headers [headers]
+  (let [s (StringBuilder.)]
+    (doseq [[k v] headers]
+      (doto s
+        (.append k) (.append ":") (.append v) (.append "\n")))
+    (.toString s)))
+
+(defn- content-sha256
+  "Returns content sha256 given payload"
+  [payload]
+  (cond
+    (= payload UNSIGNED_PAYLOAD) UNSIGNED_PAYLOAD
+    (nil? payload) EMPTY_SHA256
+    :else (sha-256 (to-utf8 payload))))
+
+(defn- canonical-request
+  "Returns canonical request as string"
+  [{:keys [method uri query payload headers]}]
+  (str
+   method \newline
+   (normalize-uri uri) \newline
+   (query->string query) \newline
+   (stringify-headers headers)   \newline
+   (str/join ";" (keys headers)) \newline
+   (or (get headers "x-amz-content-sha256")
+       (content-sha256 payload))))
+
+;; ------------------
+;; AWS authentication
+
+(defn- signature
+  "Returns hmac-256 signature given credential scope, secret and string-to-sign"
+  [{:keys [secret-key short-timestamp region service string-to-sign]}]
+  (-> (signing-key secret-key short-timestamp region service)
+      (hmac-256 string-to-sign)
+      (as-hex-str)))
+
+(defn- string-to-sign
+  "Returns string to sign given data of request-to-sign"
+  [{:keys [timestamp method uri query payload short-timestamp region service headers debug]}]
+  (let [cannonical-request-str (canonical-request
+                                {:method method :uri uri :query query
+                                 :payload payload :headers headers})]
+    (when debug
+      (println "Cannonical request")
+      (println cannonical-request-str))
+    (str
+     "AWS4-HMAC-SHA256\n"
+     timestamp "\n"
+     short-timestamp "/" region "/" service "/aws4_request" "\n"
+     (sha-256 (to-utf8 cannonical-request-str)))))
+
+(defn- amz-credential [{:keys [access-key short-date region service]}]
+  (str access-key "/" short-date "/" region "/" service "/aws4_request"))
+
+(defn generate-signature
```

Comment:
The aws-sign folks _do_ have a pretty serious test suite: 

https://github.com/joseferben/clj-aws-sign/tree/master/resources/suite

We could import those too, but was a bit conflicted, as it's a lot of files